### ISSUE TITLE
Feature Request: Mid-Training k-Fold Cross-Validation in GEPA

### ISSUE THREAD

**Issue by nickarafyllis** [gepa-ai/gepa#133]
### Summary
Add k-fold cross-validation support integrated into the GEPA optimization loop to provide robust evaluation during training, not just at the end. This ensures the optimizer selects candidates based on metrics that generalize better, rather than overfitting to a single validation set.

### Motivation
Currently, GEPA evaluates candidates using a single train/validation split. This can lead to misleading signals mid-training:

- A candidate might appear to perform very well on the validation set (e.g., 90%) but generalize poorly to unseen data (e.g., 80%).

- The optimizer may overfit to the validation set, making the search less reliable.

**Comment by LakshyAAAgrawal** [comment]
@nickarafyllis, thank you so much for the issue. Through #100, we recently added support for creating new val_evaluation_policy. While the default "full" val_evaluation_policy suffers from exactly the issue you highlight, a different val_evaluation_policy could choose when and how different data points are used for evaluation and candidate selection. Could you consider proposing new val_evaluation_policies that address this issue? I would be happy to integrate it as an option to select alongside full.

**Comment by dabblingfrancis** [comment]
@nickarafyllis came here to look for this kind of feature request, I think it's a great idea. 

I have 400 examples with a 60/20/20 train/val/test split. Use-case is classification (4 classes). I optimized two models which almost 95% accuracy on the validation set (10k rollouts, reflection_minibatch_size=20);  the first one got a similar performance on the test set but the second got <85%, which was less than the unoptimized program! 

I then tested with bootstrap sampling on the test set, to make sure it's not sure due to sampling variability or LM stochasticity, and the conclusion was the same.

**Comment by LakshyAAAgrawal** [comment]
@dabblingfrancis, thanks a lot for sharing this! I think this is a very useful feature. Do you think you would be able to implement any val_evaluation_policy to achieve this?

**Comment by dabblingfrancis** [comment]
@LakshyAAAgrawal I could look into it. 
The val_evaluation_policies looks interesting. Perhaps I could split the validation set into "candidate selection" and "evaluation", with a different split each iteration ("k-fold rotation" -> every k iterations we revisit the same split).  Not sure if this is close to what you both have in mind (@nickarafyllis)
